hail/python/hail/expr/aggregators/aggregators.py

```diff
@@ -90,91 +95,165 @@ def make_seq_op(agg):
-
-        ir = agg_expr._ir
-        agg_sig = ir.agg_sig
-        a = ir.a
-
-        new_agg_sig = AggSignature(f'Keyed({agg_sig.op})',
-                                    agg_sig.ctor_arg_types,
-                                    agg_sig.initop_arg_types,
-                                    [group.dtype] + agg_sig.seqop_arg_types)
-
-        def rewrite_a(ir):
-            if isinstance(ir, SeqOp):
-                return SeqOp(ir.i, [group._ir] + ir.args, new_agg_sig)
+    @typecheck_method(f=func_spec(1, expr_any),
+                      array_agg_expr=expr_oneof(expr_array(), expr_set()))
+    def explode(self, f, array_agg_expr):
+        if len(array_agg_expr._ir.search(lambda n: isinstance(n, BaseApplyAggOp))) != 0:
+            raise ExpressionException("'agg.explode' does not support an already-aggregated expression as the argument to 'collection'")
+
+        if isinstance(array_agg_expr.dtype, tset):
+            array_agg_expr = hl.array(array_agg_expr)
+        elt = array_agg_expr.dtype.element_type
+        var = Env.get_uid()
+        ref = construct_expr(Ref(var, elt), elt, array_agg_expr._indices)
+        self._agg_bindings.add(var)
+        aggregated = f(ref)
+        self._agg_bindings.remove(var)
+
+        def rewrite(node):
+            if isinstance(node, BaseApplyAggOp):
+                self.check_scan_agg_compatibility("explode", node)
+                old_seq_op = node.a
+                new_seq_op = ArrayFor(array_agg_expr._ir, var, old_seq_op)
+                init_ops = node.init_op_args if node.init_op_args else []
+                return node.copy(new_seq_op, *node.constructor_args, *init_ops)
+
+            elif isinstance(node, Ref) and node.name == var:
+                raise ExpressionException("'{}.explode' can't reference collection element outside of aggregation".format(self.correct_prefix()))
+            else:
+                return node.map_ir(rewrite)
+
+        aggregations = hl.utils.LinkedList(Aggregation)
+        if not self._as_scan:
+            aggregations = aggregations.push(Aggregation(aggregated))
+
+        return construct_expr(rewrite(aggregated._ir),
+                              aggregated.dtype,
+                              aggregated._indices,
+                              aggregations)
+
+    @typecheck_method(condition=expr_bool,
+                      aggregation=agg_expr(expr_any))
+    def filter(self, condition, aggregation):
+        if len(condition._ir.search(lambda n: isinstance(n, BaseApplyAggOp))) != 0:
+            raise ExpressionException("'agg.filter' does not support an already-aggregated expression as the argument to 'condition'")
+
+        unify_all(condition, aggregation)
+
+        def rewrite(node):
+            if isinstance(node, BaseApplyAggOp):
+                self.check_scan_agg_compatibility("filter", node)
+                old_seq_op = node.a
+                new_seq_op = If(condition._ir, old_seq_op, Begin([]))
```

Comment:
OK, so this is incorrect if the condition contains randomness.  I think the only way to do this correctly is with the IR changes.  While this PR looks good, I say we build the IR changes on this and then merge the whole thing in when it is ready.  How long do you think the IR changes will take?